src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -742,22 +742,24 @@ public void Init()
                 _sequenceNumber = 0;
             }
 
-            // IsReady returns the current _sequenceNumber, which must be passed to StartAsyncOperation below.
-            public bool IsReady(SocketAsyncContext context, out int observedSequenceNumber)
+            // IsReady returns whether an operation can be executed immediately without queueing it.
+            // When IsReady returns true, the observedSequenceNumber is set.
+            // This ensures the sequence number is read, before executing the operation.
+            // When the operation can't complete, StartAsyncOperation compares against the current
+            // sequence number to see if the operation should be tried again.
+            public bool IsReady(SocketAsyncContext context, out int? observedSequenceNumber)
             {
-                using (Lock())
```

Comment:
What's the rationale for why it's ok to not take the lock?

Reply:
The return value is still based on `_state`. Doing `Volatile.Read` will give the same result.

The role of the `observedSequenceNumber` is to ensure we don't miss events that require retrying.

By doing `Volatile.Read` before the op (which can't re-order), we can compare in `StartAsyncOperation`.

Without the `Lock` it is possible we observe the sequence number increment before state becomes ready. In that case `observedSequenceNumber` will be `null`, and `StartAsyncOperation` will retry the operation.

Reply:
Thanks.  Can you include the relevant details in a comment?